tests: Add target support to RepositorySimulator

jku opened the pull request:
* Add very simple targets support into simulator
* Add documentation for the simulator
* Add an example targets test

This might need to be tweaked and/or extended as we add tests but the
implementation should give a good indication of how to extend it.

As an example, non-consistent targets (or metadata) are not yet supported, but
making fetch() check for the consistent_snapshot state and respond
accordingly should be easy.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

---

This should be useful to properly test consistent targets (#1576 ) and then implement good test coverage for updater in general.

CC @MVrachev, marking you a reviewer as I know you might want to use it. If you have any improvement ideas let me know. A potential one is a way for simulator to verify targets client has downloaded (a helper, so tests could say _"assert that my target <file path> is correct"_), but I figured we'd first write some tests and see where we duplicate code...

Commit:
tests: Add target support to RepositorySimulator

* Add very simple targets support into simulator
* Add documentation for the simulator
* Add an example targets test

This might need to be tweaked and/or extended as we add tests but the
implementation should give a good indication of how to extend it.

As an example, non-consistent targets are not yet supported, but
making fetch() check for the consistent_snapshot state and respond
accordingly should be easy.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

MVrachev reviewed (approved):
I tried it locally and it looks good to me.
For now, I don't have any concrete suggestions.

MVrachev commented on tests/test_updater_with_simulator.py:
```diff
@@ -71,6 +79,33 @@ def test_refresh(self):
+
+        # target now exists, is not in cache yet
+        updater = self._run_refresh()
+        info = updater.get_one_valid_targetinfo("file")
```
`info` sounds a little ambiguous.
Maybe `file_info`?
So when we add more targets they will be `file2`, `file3` and when we get their info with `get_one_valid_targetinfo()` their corresponding variables will be `file2_info`, `file3_info`, etc.

jku replied:
yeah... I'm trying to avoid long lines in test code but you may be right: the issue in these lines is `get_one_valid_targetinfo` and `self.assertEquals` not the variable name

sechkova commented on tests/repository_simulator.py:
```diff
@@ -136,10 +176,30 @@ def fetch(self, url: str) -> Iterator[bytes]:
             raise FetcherHTTPError(f"Unknown path '{spliturl.path}'", 404)
 
+    def _fetch_target(self, target_path: str, hash: Optional[str]) -> bytes:
+        """Return data for 'target_path', checking 'hash' if it is given"""
```
Can you please mention that hash will be requested if consistent snapshot -> consistent targets' naming is used? Took me some time to understand why the hash check is done here. Although when you see metadata and version you can get the idea ... so maybe it's my problem only :)

jku replied:
added some docstrings and a NotImplementedError for `not consistent_snapshot` to make this clearer

Commit:
tests: Improve the docs on RepositorySimulator

The handling of consistent snapshot was not very clear: try to make
it more obvious what is supported and what is not.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented:
Modified based on feedback, rebased on develop
